extensions/amp-ad-network-doubleclick-impl/0.1/doubleclick-a4a-config.js

```diff
@@ -77,55 +83,111 @@ export const BETA_ATTRIBUTE = 'data-use-beta-a4a-implementation';
 /** @const {string} */
 export const BETA_EXPERIMENT_ID = '2077831';
 
-/**
- * @param {!Window} win
- * @param {!Element} element
- * @returns {boolean}
- */
-export function doubleclickIsA4AEnabled(win, element) {
-  if ('useSameDomainRenderingUntilDeprecated' in element.dataset ||
-      element.hasAttribute('useSameDomainRenderingUntilDeprecated') ||
-      !isGoogleAdsA4AValidEnvironment(win)) {
-    return false;
+/** @visibleForTesting */
+export class DoubleclickA4aEligibility {
+  constructor() {
+    return singleton;
+  }
+
+  /**
+   * Returns whether win supports native crypto. Is just a wrapper around
+   * supportsNativeCrypto, but this way we can mock out for testing.
+   * @param {!Window} win
+   * @return {boolean}
+   */
+  supportsCrypto(win) {
+    return supportsNativeCrypto(win);
   }
-  if (element.hasAttribute(BETA_ATTRIBUTE)) {
-    addExperimentIdToElement(BETA_EXPERIMENT_ID, element);
-    dev().info(TAG, `beta forced a4a selection ${element}`);
-    return true;
+
+  /**
+   * Returns whether we are running on the AMP CDN.
+   * @param {!Window} win
+   * @return {boolean}
+   */
+  isCdnProxy(win) {
+    const googleCdnProxyRegex =
+      /^https:\/\/([a-zA-Z0-9_-]+\.)?cdn\.ampproject\.org/;
+    return !!googleCdnProxyRegex.test(win.location.origin);
+  }
+
+  /** Whether Fast Fetch is enabled
+   * @param {!Window} win
+   * @param {!Element} element
+   * @return {boolean}
+   */
+  isA4aEnabled(win, element) {
+    let experimentId;
+    if ('useSameDomainRenderingUntilDeprecated' in element.dataset ||
+        element.hasAttribute('useSameDomainRenderingUntilDeprecated') ||
+        !this.supportsCrypto(win)) {
+      return false;
+    }
+    if (!this.isCdnProxy(win)) {
+      experimentId = this.selectExperiment(win, element, [
+        DOUBLECLICK_EXPERIMENT_FEATURE.CANONICAL_CONTROL,
+        DOUBLECLICK_EXPERIMENT_FEATURE.CANONICAL_EXPERIMENT,
+      ], DFP_CANONICAL_FF_EXPERIMENT_NAME);
+      if (experimentId) {
```

Comment:
You could just have an else:
let experimentName = DFP_CANONICAL_FF_EXPERIMENT_NAME;
if (!this.isCdnProxy(win)) {
    experimentId = <select into canonical exp>
} else {
    experimentName = DOUBLECLICK_A4A_EXPERIMENT_NAME;
    if (element.hasAttribute...
}
if (experimentId) {
       addExperimentIdToElement(experimentId, element);
       forceExperimentBranch(win, experimentName, experimentId);
    }
    return ![DOUBLECLICK_EXPERIMENT_FEATURE.CANONICAL_CONTROL,
      DOUBLECLICK_EXPERIMENT_FEATURE.HOLDBACK_EXTERNAL,
       DOUBLECLICK_EXPERIMENT_FEATURE.HOLDBACK_INTERNAL,
       DOUBLECLICK_EXPERIMENT_FEATURE.SFG_CONTROL_ID,
       DOUBLECLICK_EXPERIMENT_FEATURE.SFG_EXP_ID].includes(experimentId);

Reply:
done